### PR TITLE
Added Filename Filter

### PR DESCRIPTION
**Pull request by Mihaeru**
Added a textbox which allows filtering the filenames by the entered string.

This is a proposal for a functionality which I find very useful. Would be great if you could either just use it or implement it in any other way. I was also thinking about adding a button for initiating the filter and for clearing it, if FilterText is not empty. Finally I liked it more to start filtering the filenames immediately when I start typing and so also skipped the effort for adding a button to clear the textbox content.

### CODE REVIEW

**Comment by Ant-f** [comment]
Hi Mihaeru

Thanks for your interest in this project! Your feature idea is a good one. I've had a look at your pull request, and had some thoughts. In this, a new `IsFilterMatch` property is being created, and I'm thinking that this has the possibility somehow of having filter results that do not match the filter criteria.

Instead of adding `IsFilterMatch`, it might instead be possible to add to the filtering criteria used for creating `ActiveDocumentMetadata` and `PinnedDocumentMetadata` themselves:

https://github.com/Ant-f/WorkingFilesList/blob/master/Solution/WorkingFilesList.ToolWindow/ViewModel/DocumentMetadataManager.cs#L79

https://github.com/Ant-f/WorkingFilesList/blob/master/Solution/WorkingFilesList.ToolWindow/ViewModel/DocumentMetadataManager.cs#L104

This would mean that in the setter of `FilterString`, you'd only need to call `Refresh` on the two views.

Would you also be able to add some unit tests to test that they are being correctly filtered? There are some similar tests towards the end of `DocumentMetadataManagerTests`

**Comment by Ant-f** [comment]
If you're interested in making the changes yourself, would you be able to change your pull request to base off the newly created `features-filename-filter` branch?

**Comment by Mihaeru** [comment]
Hi Anthony,

thanks for you feedback and approval for the feature idea.
> filtering criteria used for creating ActiveDocumentMetadata and PinnedDocumentMetadata themselves

Ok. I will check the proposed functions and will add the filter there.

> I'm thinking that this has the possibility somehow of having filter results that do not match the filter criteria.

Indeed that's the case. Newly opened documents are not filtered but always displayed until you modify the filter again.

> Would you also be able to add some unit tests to test that they are being correctly filtered?

Yep. I will add them. Don't know when I can finish that as I'm currently a bit busy but as it shouldn't be too much effort, I can do it hopefully soon.

>If you're interested in making the changes yourself, would you be able to change your pull request to base off the newly created features-filename-filter branch?

Ok. I will try.

Thanks and best regards,
Michael

**Comment by Mihaeru** [comment]
Hi Anthony,

I modified the code like you proposed. Seems to work well. Unfortunately, I haven't been able so far to provide a working unit test. I can create documents and apply the filter string but how can I check if a document would be filtered or not?

**Comment by Ant-f** [comment]
Hi Michael

I would suggest that each collection view is tested separately. For each view, it seems like a good idea to test that:

 * an empty filter string returns everything
 * a filter string that matches some documents returns only those documents
 * a filter string that matches no documents return an empty view
 * any other scenario/edge case that you can think of

I'm thinking a test similar to this one

https://github.com/Ant-f/WorkingFilesList/blob/master/Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs#L2236

where you'd add two documents with different display names to the source collection.

In the _Act_ phase, you could set the filter string; this looks like it will refresh the views within the `DocumentMetadataManager`.

You could then create a assertions of the number of items you expect there to be in that view, and that they are what are expected.

By testing each view separately, there should be at least six tests: at least three for each view.

I would also suggest a test to assert that when creating a `DocumentMetadataManager`, its filter string is an empty string by default.

**Comment by Mihaeru** [comment]
Hi Anthony,

thanks for your support. The idea what to test is clear and basically also the idea how to do it. As I wrote before, I just don't know how to check which document is finally displayed and which is filtered out in the view.
When checking e.g. manager.ActiveDocumentMetadata, all documents are included and the filter is applied but how can I verify if it is really effective in the view:
> You could then create a assertions of the number of items you expect there to be in that view,

How to do this?

**Comment by Ant-f** [comment]
> When checking e.g. manager.ActiveDocumentMetadata, all documents are included and the filter is applied but how can I verify if it is really effective in the view:

I would expect everything to be there in `manager.ActiveDocumentMetadata.SourceCollection`: that's the underlying collection of all data. `manager.ActiveDocumentMetadata` is a filtered view and I'm surprised that it's in there if the filter string has been applied. It being there would imply that it would also be there when running: that's one of the views that the list is bound to.

If I understand correctly, setting the filter string refreshes the two views. When the views refresh, the filter functions are called (if I remember correctly). Can you check whether a breakpoint set at

https://github.com/Ant-f/WorkingFilesList/blob/master/Solution/WorkingFilesList.ToolWindow/ViewModel/DocumentMetadataManager.cs#L77

is called after setting the filter string?

> 
> You could then create a assertions of the number of items you expect there to be in that view,
> 
> How to do this?

Let's assume you have two items, and your filter string is empty: you expect both items to be there. You would find how many items there are in that view with something like

`var count = manager.ActiveDocumentMetadata.Cast<DocumentMetadata>().Count();`

and then

`Assert.AreEqual(2, count);`

If you only expect one item, you can get it with

`var metadata = manager.ActiveDocumentMetadata.Cast<DocumentMetadata>().Single();`

This returns you the item and asserts that there is only one in the view. You can then do an assertion on the display/file name.

**Comment by Mihaeru** [comment]
Hi Anthony,

I added all test cases, which came into my mind. I'm really starting to like Moq :) I never worked with that before.
Please check and let me know your opinion.
Thanks again for your support and, of course, for this great and useful extension!

Best regards,
Mihaeru

**Comment by Ant-f** [comment]
Hi Michael

Sorry I haven’t had a chance to properly look over your pull request yet. I’ve been a little busy lately, but I should be able to do so some time soon. In the meantime, would you be so kind as to include your sign off [here](https://github.com/Ant-f/WorkingFilesList/blob/features-filename-filter/Contributors.txt) as part of your pull request (you may need to pull the latest)?

Thanks

Anthony

**Comment by Mihaeru** [comment]
Hi Anthony,

no problem. Being busy is not an unknown situation for me :)
If you have any concerns or proposals regarding the unit tests or whatever, please let me know.

I signed the contributors list and also tried to integrate it to the pull request but as I created my changes on master branch and the new file is on your feature branch, there was a conflict when trying to commit. I hope I solved it correctly - at least I assumed that I cannot accidentally modify your original project.

Best regards,
Michael

**Comment by Ant-f** [comment]
Thanks for adding the tests. I’ve managed to have a look and would like to suggest a few changes. As `DocumentMetadataManagerTests` has many tests in it, I think it would be good to group the new tests together; I’d like to suggest creating an inner class to put all the new tests and related methods into – something like:

```
public class DocumentMetadataManagerTests
{
  // Existing test { ... }

  // Existing test { ... }

  public class FilterStringTests
  {
    // New FilterString test { ... }

    // New FilterString test { ... }
  }
}
```

File: Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs [hunk 1]
```diff
@@ -2294,5 +2294,236 @@ public void ClearRemovesAllMetadata()
             Assert.IsEmpty(manager.PinnedDocumentMetadata);
             Assert.IsEmpty(collection);
         }
+
+        public IDocumentMetadataManager PrepareDocumentsForFilterTest(params string[] docNames)
```
Comment (Ant-f) [inline Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs]: This is only used by the new tests and could be `private`

File: Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs [hunk 2]
```diff
@@ -2294,5 +2294,236 @@ public void ClearRemovesAllMetadata()
...
+            Assert.IsFalse(manager.PinnedDocumentMetadata.Contains(document2));
+
+            var document3 = collection.Single(m => m.FullName == document3Name);
+            Assert.IsFalse(manager.ActiveDocumentMetadata.Contains(document3));
+            Assert.IsFalse(manager.PinnedDocumentMetadata.Contains(document3));
+        }
+
+
+        [Test]
+        public void FilterIsEmpytShowingAllDocuments()
```
Comment (Ant-f) [inline Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs]: Small typo: `Empyt`

Would also suggest turning this into a parameterised test, i.e. `[TestCase()]`, with cases for `""`, `null`, and whitespace

File: Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs [hunk 3]
```diff
@@ -2294,5 +2294,236 @@ public void ClearRemovesAllMetadata()
...
+        }
+
+        [Test]
+        public void FilterMatchesTwoDocuments()
+        {
+            // Arrange
+            const string document1Name = "AAADocument1";
+            const string document2Name = "BBBDocument2";
+            const string document3Name = "AABDocument3";
+
```
Comment (Ant-f) [inline Solution/WorkingFilesList.ToolWindow.Test/ViewModel/DocumentMetadataManagerTests.cs]: As all of the tests use three or four document names that are fairly similar, I’d suggest pulling these out as `const string`s into the new inner class; something like
```
private const string AaaDocument1 = "AAADocument1";
```

**Comment by Mihaeru** [comment]
Sorry for the delay. This time I was quite busy and modified the code whenever I had some minutes to spare. Unfortunately I saw your Pull request a bit late and already started to extract the unit tests to an inner class as well.
Anyhow, I confirmed your pull request for the unit tests and added the remaining review items plus an additional check for covering case insensitive match.

**Comment by Ant-f** [comment]
Hi Michael

Thanks for updating the tests; I've merged in your pull request.

On having another look, I saw an opportunity to convert the tests into test cases, meaning that I could also add similar tests for checking that the Active and Pinned metadata collections can operate independently. I also made a few other changes to help the search bar fit in a bit more with the look and feel of Visual Studio.

I've built a release candidate that I plan to run for a little while for testing. If you're eager to use the search bar and would like to run it, please feel free to do so:

[WorkingFilesList_1_5_0.zip](https://github.com/Ant-f/WorkingFilesList/files/3769298/WorkingFilesList_1_5_0.zip)

**Comment by Mihaeru** [comment]
Hi Anthony,
sorry for the long delay. I just returned from vacation.
The new version looks and feels really good. Initially I also implemented an icon for clearing but it was just a prototype, implemented more like a hack. and then I though it's not worth the effort. But yes, I have to agree that it looks better now and is convenient. Especially with the search icon the view is really improved.

Thanks again for agreeing to this feature and improving it.
